[MDS-4613] Allow creation / editing of a TSF Qualified Person through MineSpace

simensma-fresh opened the pull request:
## Objective 

[MDS-4613](https://bcmines.atlassian.net/browse/MDS-4613)
[MDS-4725](https://bcmines.atlassian.net/browse/MDS-4725)
[MDS-4726](https://bcmines.atlassian.net/browse/MDS-4726)

_Why are you making this change? Provide a short explanation and/or screenshots_

Added support for MineSpace users to assign/create update Qualified Persons to a TSF. Fixed various bugs in the contact creation modal, TSF flow.

![image](https://user-images.githubusercontent.com/66635118/192653290-85dbfc81-e8f6-4f2c-ae57-502ecdd37300.png)
![image](https://user-images.githubusercontent.com/66635118/192653315-40b09010-dcda-4bfe-ace9-e2ad54ef99ae.png)

simensma-fresh committed:
[MDS-4613] Added support for MS managed Qualified Persons

simensma-fresh committed:
[MDS-4613] Added Qualified Person tab to tailings page

simensma-fresh committed:
MDS-4613 Added filtering by assignment type in Add Contact modal

matbusby-fw commented on services/core-api/app/api/mines/tailings/models/tailings.py:
```diff
@@ -63,11 +63,25 @@ class MineTailingsStorageFacility(AuditMixin, Base):
+    @hybrid_property
+    def qualified_person(self):
+        if self.qualified_persons:
+            return self.qualified_persons[0]
```
Are we only allowing a single qualified person?  I thought there could be more than one, but I may be wrong about that.

simensma-fresh replied:
Currently, a mine can have multiple Qualified Persons since they are not tied to a TSF at all, but my understanding is that a TSF will only have one QP going forward

matbusby-fw replied:
Maybe you're right.  The designs certainly support that look.  I just had a vague memory of discussing multiple TSF QPs, but I think that's either not what's needed or at least not MVP.

matbusby-fw commented on services/core-api/app/api/parties/party_appt/models/mine_party_appt.py:
```diff
@@ -64,9 +64,9 @@ def validate_permit():
+            if self.mine_party_appt_type_code in TSF_ALLOWED_CONTACT_TYPES:
                 if not self.mine_tailings_storage_facility:
-                    raise AssertionError('The associated TSF is required for Engineer of Records.')
+                    raise AssertionError('The associated TSF is required for Engineer of Records or Qualified Persons.')
```
Probably not a big deal, but since Qualified Persons can exist for other Mine contacts, maybe we could change the last part to something like 'TSF related qualified persons'?

matbusby-fw commented on services/minespace-web/src/components/common/SteppedForm.js:
```diff
@@ -48,7 +48,8 @@ const SteppedForm = (props) => {
   const handleNextClick = (evt, tab) => {
     evt.preventDefault();
-    handleTabClick(tab);
+    handleSaveData(null, tab);
```
Hmmm, we'll need to think how we want to handle saves on tab changes.  I think this is going to trigger a double save for 'Basic Information' and 'EOR' since those saves are triggered within `SteppedForm`

simensma-fresh replied:
Good point! My opinion would be that saves should only happen when you hit "next" and not when you switch tabs, if anything we could reset the form with a "are you sure you want to discard your current changes?" prompt if you switch tabs without hitting next. 

Will check in with Roop/Raj on this and update it across the board

matbusby-fw replied:
Yeah, that would work for me!

matbusby-fw reviewed (changes requested):
Looks great!  Just a couple of questions.  I added a couple of questions.  I had thought we were doing multiple QPs, but it's quite possible that I have that wrong.

simensma-fresh committed:
MDS-4613 Fixed eor + QP errors

simensma-fresh committed:
MDS-4613 Fixed snapshots

simensma-fresh commented:
@matbusby-fw I've updated the PR with some of the changes if you want to take another look
- All tabs except the basic-info tab are now disabled until the TSF exists instead of basing it on form errors
- Saving/creating of TSF/EOR/QP now only happens when you hit "next", not when you click on the tabs
- The state now fully reloads after you add a new EOR/QP, to avoid any inconsistencies (e.g. EOR document not showing up when you hit back button until you refresh the page). 
- Fixed the related issues in MDS-4725 / MDS-4726

matbusby-fw reviewed (approved):
Awesome!  Looks great!
